src/cunumeric/utilities/dispatch.h

```diff
@@ -0,0 +1,192 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#pragma once
+
+#include "legate.h"
+#include "cunumeric/cunumeric_c.h"
+
+namespace cunumeric {
+
+template <int DIM>
+struct inner_type_dispatch_fn {
+  template <typename TypeCode, typename Functor, typename... Fnargs>
+  constexpr decltype(auto) operator()(TypeCode code, Functor f, Fnargs&&... args)
+  {
+    switch (code) {
+      case CuNumericTypeCodes::CUNUMERIC_TYPE_POINT1: {
+        return f.template operator()<CuNumericTypeCodes::CUNUMERIC_TYPE_POINT1, DIM>(
+          std::forward<Fnargs>(args)...);
+      }
+      case CuNumericTypeCodes::CUNUMERIC_TYPE_POINT2: {
+        return f.template operator()<CuNumericTypeCodes::CUNUMERIC_TYPE_POINT2, DIM>(
+          std::forward<Fnargs>(args)...);
+      }
+      case CuNumericTypeCodes::CUNUMERIC_TYPE_POINT3: {
+        return f.template operator()<CuNumericTypeCodes::CUNUMERIC_TYPE_POINT3, DIM>(
+          std::forward<Fnargs>(args)...);
+      }
+      case CuNumericTypeCodes::CUNUMERIC_TYPE_POINT4: {
+        return f.template operator()<CuNumericTypeCodes::CUNUMERIC_TYPE_POINT4, DIM>(
+          std::forward<Fnargs>(args)...);
+      }
+      case CuNumericTypeCodes::CUNUMERIC_TYPE_POINT5: {
+        return f.template operator()<CuNumericTypeCodes::CUNUMERIC_TYPE_POINT5, DIM>(
+          std::forward<Fnargs>(args)...);
+      }
+      case CuNumericTypeCodes::CUNUMERIC_TYPE_POINT6: {
+        return f.template operator()<CuNumericTypeCodes::CUNUMERIC_TYPE_POINT6, DIM>(
+          std::forward<Fnargs>(args)...);
+      }
+      case CuNumericTypeCodes::CUNUMERIC_TYPE_POINT7: {
+        return f.template operator()<CuNumericTypeCodes::CUNUMERIC_TYPE_POINT7, DIM>(
+          std::forward<Fnargs>(args)...);
+      }
+      case CuNumericTypeCodes::CUNUMERIC_TYPE_POINT8: {
+        return f.template operator()<CuNumericTypeCodes::CUNUMERIC_TYPE_POINT8, DIM>(
+          std::forward<Fnargs>(args)...);
+      }
+      case CuNumericTypeCodes::CUNUMERIC_TYPE_POINT9: {
+        return f.template operator()<CuNumericTypeCodes::CUNUMERIC_TYPE_POINT9, DIM>(
+          std::forward<Fnargs>(args)...);
+      }
+      default: {
+        return legate::inner_type_dispatch_fn<DIM>{}(code, f, std::forward<Fnargs>(args)...);
+      }
+    }
+    assert(false);
+    return f.template operator()<CuNumericTypeCodes::CUNUMERIC_TYPE_POINT1, DIM>(
+      std::forward<Fnargs>(args)...);
+  }
+};
+
+template <typename TypeCode, typename Functor, typename... Fnargs>
+constexpr decltype(auto) double_dispatch(int dim, TypeCode code, Functor f, Fnargs&&... args)
+{
+  switch (dim) {
+#if LEGION_MAX_DIM >= 1
+    case 1: {
+      return cunumeric::inner_type_dispatch_fn<1>{}(code, f, std::forward<Fnargs>(args)...);
+    }
+#endif
+#if LEGION_MAX_DIM >= 2
+    case 2: {
+      return inner_type_dispatch_fn<2>{}(code, f, std::forward<Fnargs>(args)...);
```

Comment:
why is this not `cunumeric::inner_type_dispatch_fn`? (same question to other cases)

Reply:
Yes, it should have cunumeric:: in front.